File: Nori/Panel/HintBarModel.swift
import Foundation

/// What the hint bar shows for the current modifier state. Pure, so it is tested as a table
/// and can never disagree with `ActionGrammar`.
enum HintBarModel {
    struct Chip: Equatable, Sendable {
        var key: String
        var verb: String
        /// True for the "enable pasting" chip that opens System Settings when clicked.
        var isWarning = false
    }

    struct Input: Sendable {
        var bits: ActionGrammar.Bits
        var accessibilityTrusted: Bool
        var cycleMode: Bool
        /// The hotkey's modifier glyphs (for cycle mode), e.g. "⇧⌘".
        var hotkeyModifiers: String
        var selectedKind: ClipKind?
        /// False in the empty and no-match states, where ↩ / Space / ⌘P / ⌘⌫ have no card to act on.
        var hasSelection: Bool
        /// Masked secrets cannot be previewed, pinned, or pasted with keep-open.
        var isSensitive = false
        /// The trimmed search text; with no selection, ↩ pastes it as plain text.
        var query = ""
    }

    static func chips(_ input: Input) -> [Chip] {
        let trusted = input.accessibilityTrusted
        if input.cycleMode {
            return [
                Chip(key: String(localized: "Release \(input.hotkeyModifiers)"),
                     verb: trusted ? String(localized: "to paste") : String(localized: "to copy")),
                Chip(key: "↑↓", verb: String(localized: "Move")),
                Chip(key: String(localized: "Esc"), verb: String(localized: "Cancel")),
            ]
        }

        guard input.hasSelection else {
            // Nothing to paste: only the typed query can go somewhere.
            guard !input.query.isEmpty else { return [] }
            let verb = trusted
                ? String(localized: "Paste “\(input.query)” as text")
                : String(localized: "Copy “\(input.query)” as text")
            return [Chip(key: "↩", verb: verb), Chip(key: "⌃U", verb: String(localized: "Clear"))]
        }

        let caps = ActionGrammar.Capabilities(accessibilityTrusted: trusted)
        let sensitive = input.isSensitive
        var chips: [Chip] = []

        let bits = input.bits
        if bits.isEmpty {
            if trusted {
                chips.append(Chip(key: "↩", verb: String(localized: "Paste")))
            } else {
                chips.append(Chip(key: "↩", verb: String(localized: "Copy · Enable pasting →"), isWarning: true))
            }
            // Resting state stays short on purpose; ⌥ and ⌘ variants appear while those keys are held.
            chips.append(Chip(key: "⇧↩", verb: String(localized: "Plain")))
            if !sensitive {
                chips.append(Chip(key: String(localized: "Space"), verb: String(localized: "Preview")))
                chips.append(Chip(key: "⌘P", verb: String(localized: "Pin")))
            }
            chips.append(Chip(key: "⌘⌫", verb: String(localized: "Delete")))
        } else if bits == [.copyOnly] {
            chips.append(Chip(key: "⌘1–9", verb: trusted ? String(localized: "Paste item") : String(localized: "Copy item")))
            chips.append(Chip(key: "⌘↩", verb: String(localized: "Copy")))
            if !sensitive {
                chips.append(Chip(key: "⌘P", verb: String(localized: "Pin")))
                chips.append(Chip(key: "⌘Y", verb: String(localized: "Preview")))
            }
            if input.selectedKind == .link || input.selectedKind == .file || input.selectedKind == .image {
                chips.append(Chip(key: "⌘O", verb: String(localized: "Open")))
            }
            if input.selectedKind == .file {
                chips.append(Chip(key: "⌘R", verb: String(localized: "Reveal")))
            }
            chips.append(Chip(key: "⌘⌫", verb: String(localized: "Delete")))
            chips.append(Chip(key: "⌘⇧⌫", verb: String(localized: "Clear…")))
        } else if bits == [.plain] {
            chips.append(Chip(key: "⇧↩", verb: trusted ? String(localized: "Paste as plain text") : String(localized: "Copy as plain text")))
            chips.append(Chip(key: String(localized: "⇧-click"), verb: String(localized: "Same")))
            chips.append(Chip(key: "⇧⌘1–9", verb: trusted ? String(localized: "Paste item as plain text") : String(localized: "Copy item as plain text")))
        } else if bits == [.keepOpen] {
            if !sensitive {
                chips.append(Chip(key: "⌥↩", verb: trusted ? String(localized: "Paste and keep Nori open") : String(localized: "Copy and keep Nori open")))
                chips.append(Chip(key: String(localized: "⌥-click"), verb: String(localized: "Same")))
            }
            chips.append(Chip(key: "⌥⌘1–9", verb: trusted ? String(localized: "Paste item, keep open") : String(localized: "Copy item, keep open")))
        } else if bits == [.plain, .copyOnly] {
            chips.append(Chip(key: "⇧⌘↩", verb: String(localized: "Copy as plain text")))
            chips.append(Chip(key: "⇧⌘1–9", verb: trusted ? String(localized: "Paste item as plain text") : String(localized: "Copy item as plain text")))
        } else if bits == [.keepOpen, .copyOnly] {
            if !sensitive {
                chips.append(Chip(key: "⌥⌘↩", verb: String(localized: "Copy, keep open")))
            }
            chips.append(Chip(key: "⌥⌘1–9", verb: trusted ? String(localized: "Paste item, keep open") : String(localized: "Copy item, keep open")))
        } else {
            // ⇧⌥ with or without ⌘: show the stacked result.
            if !sensitive {
                let action = ActionGrammar.resolve(.returnKey, bits, caps)
                chips.append(Chip(key: "\(bits.glyphs)↩", verb: ActionGrammar.verb(for: action)))
            }
            let numberAction = ActionGrammar.resolve(.number(1), bits, caps)
            let numberVerb = ActionGrammar.verb(for: numberAction)
            chips.append(Chip(key: "\(bits.glyphs.replacingOccurrences(of: "⌘", with: ""))⌘1–9", verb: String(localized: "\(numberVerb) (item)")))
        }
        return chips
    }
}

File: windows/Nori.Core/HintBarModel.cs
namespace Nori.Core;

/// <summary>
/// What the hint bar shows for the current modifier state. Pure, so it is tested as a table and can
/// never disagree with <see cref="ActionGrammar"/>. Keys are Windows key names; verbs are resource keys.
/// </summary>
public static class HintBarModel
{
    /// <param name="Key">Keycap text, e.g. "Enter", "Shift+Enter", "Ctrl+1–9".</param>
    /// <param name="VerbKey">Resource key of the verb, e.g. "Verb_Paste", "Hint_Preview".</param>
    /// <param name="IsWarning">True for the chip that explains why Enter copies instead of pasting.</param>
    public sealed record Chip(string Key, string VerbKey, bool IsWarning = false)
    {
        public override string ToString() => $"{Key} {VerbKey}";
    }

    public sealed record Input(
        ActionGrammar.Bits Bits,
        bool CanPaste,
        bool CycleMode,
        // The hotkey's modifier text (for cycle mode), e.g. "Ctrl+Shift".
        string HotkeyModifiers,
        ClipKind? SelectedKind,
        bool HasSelection);

    public const string NumberRow = "1–9";

    public static IReadOnlyList<Chip> Chips(Input input)
    {
        if (input.CycleMode)
        {
            return
            [
                new Chip($"Release {input.HotkeyModifiers}", "Hint_ToPaste"),
                new Chip("↑↓", "Hint_Move"),
                new Chip("Esc", "Hint_Cancel"),
            ];
        }

        var caps = new ActionGrammar.Capabilities(input.CanPaste);
        var itemVerb = input.CanPaste ? "Hint_PasteItem" : "Hint_CopyItem";
        var itemPlain = input.CanPaste ? "Hint_PasteItemPlain" : "Hint_CopyItemPlain";
        var itemKeepOpen = input.CanPaste ? "Hint_PasteItemKeepOpen" : "Hint_CopyItemKeepOpen";
        var bits = input.Bits;
        var chips = new List<Chip>();

        switch (bits)
        {
            case ActionGrammar.Bits.None:
                chips.Add(input.CanPaste ? new Chip("Enter", "Verb_Paste") : new Chip("Enter", "Hint_CopyNoTarget", IsWarning: true));
                chips.Add(new Chip("Shift+Enter", "Hint_Plain"));
                chips.Add(new Chip("Space", "Hint_Preview"));
                chips.Add(new Chip("Ctrl+P", "Hint_Pin"));
                chips.Add(new Chip("Del", "Hint_Delete"));
                break;
            case ActionGrammar.Bits.CopyOnly:
                chips.Add(new Chip($"Ctrl+{NumberRow}", itemVerb));
                chips.Add(new Chip("Ctrl+Enter", "Verb_Copy"));
                chips.Add(new Chip("Ctrl+P", "Hint_Pin"));
                chips.Add(new Chip("Ctrl+Y", "Hint_Preview"));
                if (input.SelectedKind is ClipKind.Link or ClipKind.File or ClipKind.Image)
                {
                    chips.Add(new Chip("Ctrl+O", "Hint_Open"));
                }
                if (input.SelectedKind == ClipKind.File)
                {
                    chips.Add(new Chip("Ctrl+R", "Hint_Reveal"));
                }
                chips.Add(new Chip("Ctrl+Backspace", "Hint_Delete"));
                chips.Add(new Chip("Ctrl+Shift+Backspace", "Hint_Clear"));
                break;
            case ActionGrammar.Bits.Plain:
                chips.Add(new Chip("Shift+Enter", input.CanPaste ? "Verb_PastePlain" : "Verb_CopyPlain"));
                chips.Add(new Chip("Shift+click", "Hint_Same"));
                chips.Add(new Chip($"Ctrl+Shift+{NumberRow}", itemPlain));
                break;
            case ActionGrammar.Bits.KeepOpen:
                chips.Add(new Chip("Alt+Enter", input.CanPaste ? "Verb_PasteKeepOpen" : "Verb_CopyKeepOpen"));
                chips.Add(new Chip("Alt+click", "Hint_Same"));
                chips.Add(new Chip($"Ctrl+Alt+{NumberRow}", itemKeepOpen));
                break;
            case ActionGrammar.Bits.Plain | ActionGrammar.Bits.CopyOnly:
                chips.Add(new Chip("Ctrl+Shift+Enter", "Verb_CopyPlain"));
                chips.Add(new Chip($"Ctrl+Shift+{NumberRow}", itemPlain));
                break;
            case ActionGrammar.Bits.KeepOpen | ActionGrammar.Bits.CopyOnly:
                chips.Add(new Chip("Ctrl+Alt+Enter", "Verb_CopyKeepOpen"));
                chips.Add(new Chip($"Ctrl+Alt+{NumberRow}", itemKeepOpen));
                break;
            default:
            {
                // Shift+Alt with or without Ctrl: show the stacked result.
                var action = ActionGrammar.Resolve(ActionGrammar.Base.Return, bits, caps);
                chips.Add(new Chip($"{ActionGrammar.ModifierPrefix(bits)}Enter", ActionGrammar.VerbKey(action)));
                var numberAction = ActionGrammar.Resolve(new ActionGrammar.Base.Number(1), bits, caps);
                var numberBits = bits | ActionGrammar.Bits.CopyOnly;
                chips.Add(new Chip($"{ActionGrammar.ModifierPrefix(numberBits)}{NumberRow}", ActionGrammar.VerbKey(numberAction) + "_Item"));
                break;
            }
        }
        return chips;
    }
}
